File: 0285-inorder-successor-in-bst/0285-inorder-successor-in-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* inorderSuccessor(TreeNode* root, TreeNode* p) {
        TreeNode *node = root,*successor = NULL; 
        while(node){
            if(node==p){
                TreeNode *temp = node->right; 
                while(temp && temp->left) temp=temp->left;
                return temp?temp:successor;
            }
            if(node->val>p->val){
                successor=node;
                node=node->left;
            }
            else {
                node=node->right;
            }
        }
        return NULL; 
    }
};

File: 0285-inorder-successor-in-bst/0285-inorder-successor-in-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderSuccessor(self, root: TreeNode, p: TreeNode, parent=None) -> Optional[TreeNode]:
        if not(root): 
            return root 
        if root.val==p.val: 
            root = root.right 
            while root and root.left: 
                root=root.left 
            return root if root else parent

        if root.val>p.val:
            return  self.inorderSuccessor(root.left,p,root)
        return self.inorderSuccessor(root.right,p,parent)
        